[flaky] pkg/channel/distributed/common/metrics: TestMetricsServer_Report

https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative-sandbox_eventing-kafka/132/pull-knative-sandbox-eventing-kafka-unit-tests/1330891148914331648
```
=== RUN   TestMetricsServer_Report
    logger.go:130: 2020-11-23T15:12:34.872Z	INFO	profiling/server.go:59	Profiling enabled: false
    logger.go:130: 2020-11-23T15:12:34.887Z	INFO	metrics/exporter.go:160	Flushing the existing exporter before setting up the new exporter.
    logger.go:130: 2020-11-23T15:12:34.896Z	INFO	metrics/prometheus_exporter.go:51	Created Opencensus Prometheus exporter with config: &{eventing-kafka eventing_kafka prometheus 5000000000 <nil> <nil>  false 9878 0.0.0.0 false   {   false}}. Start the server for Prometheus exporter.
    logger.go:130: 2020-11-23T15:12:34.896Z	INFO	metrics/exporter.go:173	Successfully updated the metrics exporter; old config: <nil>; new config &{eventing-kafka eventing_kafka prometheus 5000000000 <nil> <nil>  false 9878 0.0.0.0 false   {   false}}
    stats_reporter_test.go:103: 
        	Error Trace:	stats_reporter_test.go:103
        	Error:      	Should be true
        	Test:       	TestMetricsServer_Report
--- FAIL: TestMetricsServer_Report (0.21s)
```

--- a/cmd/channel/distributed/dispatcher/main.go
+++ b/cmd/channel/distributed/dispatcher/main.go
@@ -96,7 +96,10 @@ func main() {
 
 	// Start The Liveness And Readiness Servers
 	healthServer := dispatcherhealth.NewDispatcherHealthServer(strconv.Itoa(environment.HealthPort))
-	healthServer.Start(logger)
+	err = healthServer.Start(logger)
+	if err != nil {
+		logger.Fatal("Failed To Initialize Health Server - Terminating", zap.Error(err))
+	}
 
 	statsReporter := metrics.NewStatsReporter(logger)
 

--- a/cmd/channel/distributed/receiver/main.go
+++ b/cmd/channel/distributed/receiver/main.go
@@ -99,8 +99,10 @@ func main() {
 
 	// Start The Liveness And Readiness Servers
 	healthServer := channelhealth.NewChannelHealthServer(strconv.Itoa(environment.HealthPort))
-	healthServer.Start(logger)
-
+	err = healthServer.Start(logger)
+	if err != nil {
+		logger.Fatal("Failed To Initialize Health Server - Terminating", zap.Error(err))
+	}
 	// Initialize The KafkaChannel Lister Used To Validate Events
 	err = channel.InitializeKafkaChannelLister(ctx, *serverURL, *kubeconfig, healthServer, environment.ResyncPeriod)
 	if err != nil {

--- a/pkg/channel/distributed/common/config/observability.go
+++ b/pkg/channel/distributed/common/config/observability.go
@@ -34,9 +34,13 @@ import (
 	"knative.dev/pkg/profiling"
 )
 
-var ListenAndServeWrapper = func(srv *nethttp.Server) func() error { return srv.ListenAndServe }
-var StartWatcherWrapper = func(cmw *configmap.InformedWatcher, done <-chan struct{}) error { return cmw.Start(done) }
-var UpdateExporterWrapper = metrics.UpdateExporter
+// These wrapper functions are to facilitate minimally-invasive unit testing of the
+// InitializeObservability functionality without requiring live servers to be started.
+var (
+	ListenAndServeWrapper = func(srv *nethttp.Server) func() error { return srv.ListenAndServe }
+	StartWatcherWrapper   = func(cmw *configmap.InformedWatcher, done <-chan struct{}) error { return cmw.Start(done) }
+	UpdateExporterWrapper = metrics.UpdateExporter
+)
 
 //
 // Initialize The Specified Context With A Profiling Server (ConfigMap Watcher And HTTP Endpoint)

--- a/pkg/channel/distributed/common/config/observability.go
+++ b/pkg/channel/distributed/common/config/observability.go
@@ -21,6 +21,8 @@ import (
 	nethttp "net/http"
 	"strings"
 
+	"knative.dev/pkg/configmap"
+
 	"go.uber.org/zap"
 	"golang.org/x/sync/errgroup"
 	corev1 "k8s.io/api/core/v1"
@@ -31,6 +33,10 @@ import (
 	"knative.dev/pkg/metrics"
 	"knative.dev/pkg/profiling"
 )
+
+var ListenAndServeWrapper = func(srv *nethttp.Server) func() error { return srv.ListenAndServe }
+var StartWatcherWrapper = func(cmw *configmap.InformedWatcher, done <-chan struct{}) error { return cmw.Start(done) }
+var UpdateExporterWrapper = metrics.UpdateExporter
 
 //
 // Initialize The Specified Context With A Profiling Server (ConfigMap Watcher And HTTP Endpoint)
@@ -43,7 +49,7 @@ func InitializeObservability(ctx context.Context, logger *zap.SugaredLogger, met
 	profilingHandler := profiling.NewHandler(logger, false)
 	profilingServer := profiling.NewServer(profilingHandler)
 	eg, egCtx := errgroup.WithContext(ctx)
-	eg.Go(profilingServer.ListenAndServe)
+	eg.Go(ListenAndServeWrapper(profilingServer))
 	go func() {
 		// This will block until either a signal arrives or one of the grouped functions
 		// returns an error.
@@ -74,7 +80,7 @@ func InitializeObservability(ctx context.Context, logger *zap.SugaredLogger, met
 		metav1.GetOptions{}); err == nil {
 		cmw.Watch(metrics.ConfigMapName(),
 			func(configMap *corev1.ConfigMap) {
-				err := metrics.UpdateExporter(ctx, metrics.ExporterOptions{
+				err := UpdateExporterWrapper(ctx, metrics.ExporterOptions{
 					Domain:         metrics.Domain(),
 					Component:      strings.ReplaceAll(metricsDomain, "-", "_"),
 					ConfigMap:      configMap.Data,
@@ -91,7 +97,7 @@ func InitializeObservability(ctx context.Context, logger *zap.SugaredLogger, met
 		return err
 	}
 
-	if err := cmw.Start(ctx.Done()); err != nil {
+	if err := StartWatcherWrapper(cmw, ctx.Done()); err != nil {
 		logger.Error("Failed to start observability configuration manager", zap.Error(err))
 		return err
 	}

--- a/pkg/channel/distributed/common/config/observability_test.go
+++ b/pkg/channel/distributed/common/config/observability_test.go
@@ -18,18 +18,21 @@ package config
 
 import (
 	"context"
-	"fmt"
+	"errors"
+	"net/http"
 	"os"
 	"testing"
-	"time"
 
 	"github.com/stretchr/testify/assert"
+	"go.opencensus.io/stats/view"
+	"go.uber.org/zap"
 	corev1 "k8s.io/api/core/v1"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes/fake"
 	"knative.dev/eventing-kafka/pkg/channel/distributed/common/env"
 	commontesting "knative.dev/eventing-kafka/pkg/channel/distributed/common/testing"
 	injectionclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/configmap"
 	logtesting "knative.dev/pkg/logging/testing"
 	"knative.dev/pkg/metrics"
 	"knative.dev/pkg/system"
@@ -37,7 +40,6 @@ import (
 
 // Test The InitializeObservability() Functionality
 func TestInitializeObservability(t *testing.T) {
-
 	// Test Data
 	ctx := context.TODO()
 	metricsPort := 9877
@@ -72,20 +74,50 @@ func TestInitializeObservability(t *testing.T) {
 	// Add The Fake K8S Client To The Context (Required By InitializeObservability)
 	ctx = context.WithValue(ctx, injectionclient.Key{}, fakeK8sClient)
 
+	// Set up some mocks so that we don't need to actually start local servers to run through
+	// the core InitializeObservability functionality
+
+	ListenAndServeWrapperRef := ListenAndServeWrapper
+	defer func() { ListenAndServeWrapper = ListenAndServeWrapperRef }()
+
+	StartWatcherWrapperRef := StartWatcherWrapper
+	defer func() { StartWatcherWrapper = StartWatcherWrapperRef }()
+
+	UpdateExporterWrapperRef := UpdateExporterWrapper
+	defer func() { UpdateExporterWrapper = UpdateExporterWrapperRef }()
+
+	calledStartWatcher := false
+	calledUpdateExporter := false
+
+	ListenAndServeWrapper = func(srv *http.Server) func() error {
+		return func() error {
+			return nil
+		}
+	}
+
+	UpdateExporterWrapper = func(ctx context.Context, ops metrics.ExporterOptions, logger *zap.SugaredLogger) error {
+		calledUpdateExporter = true
+		return nil
+	}
+
+	StartWatcherWrapper = func(cmw *configmap.InformedWatcher, done <-chan struct{}) error {
+		calledStartWatcher = true
+		cmw.OnChange(tracingConfigMap)
+		return nil
+	}
+
 	// Perform The Test (Initialize The Observability Watcher)
 	err := InitializeObservability(ctx, logger, metricsDomain, metricsPort, system.Namespace())
 	assert.Nil(t, err)
 
-	// Verify that the profiling endpoint exists and responds to requests
-	assertGet(t, "http://localhost:8008/debug/pprof", 200, 404)
-	// Verify that the metrics endpoint exists and responds to requests
-	assertGet(t, fmt.Sprintf("http://localhost:%v/metrics", metricsPort), 200, 404)
-}
+	assert.True(t, calledStartWatcher)
+	assert.True(t, calledUpdateExporter)
 
-func assertGet(t *testing.T, url string, expected int, retryIfResponse int) {
-	resp, err := commontesting.RetryGet(url, 100*time.Millisecond, 20, retryIfResponse)
-	assert.Nil(t, err)
-	assert.Equal(t, expected, resp.StatusCode)
-	err = resp.Body.Close()
-	assert.Nil(t, err)
+	// Unregister the views as registering them twice causes a panic
+	view.Unregister(metrics.NewMemStatsAll().DefaultViews()...)
+	StartWatcherWrapper = func(cmw *configmap.InformedWatcher, done <-chan struct{}) error { return errors.New("failure") }
+
+	// Test error conditions from the watcher
+	err = InitializeObservability(ctx, logger, metricsDomain, metricsPort, system.Namespace())
+	assert.NotNil(t, err)
 }

--- a/pkg/channel/distributed/common/health/health.go
+++ b/pkg/channel/distributed/common/health/health.go
@@ -86,10 +86,11 @@ func (hs *Server) initializeServer(httpPort string) {
 }
 
 // Start The HTTP Server (Blocking Call)
-func (hs *Server) Start(logger *zap.Logger) {
+func (hs *Server) Start(logger *zap.Logger) error {
 	listener, err := net.Listen("tcp", ":"+hs.HttpPort)
 	if err != nil {
 		logger.Error("Server HTTP Listen Returned Error", zap.Error(err))
+		return err
 	}
 
 	// Set the HttpPort field to whatever port was actually used by the system
@@ -105,6 +106,8 @@ func (hs *Server) Start(logger *zap.Logger) {
 			logger.Info("Server HTTP Serve Returned Error", zap.Error(err)) // Info log since it could just be normal shutdown
 		}
 	}()
+
+	return nil
 }
 
 // Stop The HTTP Server Listening For Requests

--- a/pkg/channel/distributed/common/health/health_test.go
+++ b/pkg/channel/distributed/common/health/health_test.go
@@ -17,7 +17,6 @@ limitations under the License.
 package health
 
 import (
-	"fmt"
 	"io"
 	"net/http"
 	"net/http/httptest"
@@ -38,7 +37,8 @@ const (
 
 // Test Struct That Implements The HealthInterface Functions
 type testStatus struct {
-	server *Server
+	server  *Server
+	IsReady bool
 }
 
 // Mock Function That Returns The Mock Status "Alive" Flag
@@ -48,13 +48,14 @@ func (ts *testStatus) Alive() bool {
 
 // Mock Function That Returns True For Readiness Flag
 func (ts *testStatus) Ready() bool {
-	return true
+	return ts.IsReady
 }
 
 // Mock Status For Starting Health Server
 var mockStatus testStatus
 
 func getTestHealthServer() *Server {
+	mockStatus.IsReady = true
 	health := NewHealthServer("0", &mockStatus)
 	mockStatus.server = health
 	return health
@@ -130,33 +131,23 @@ func TestHealthHandler(t *testing.T) {
 	// Verify that the shutdown process sets liveness to false
 	health.Shutdown()
 	getEventToHandler(t, health.HandleLiveness, livenessPath, http.StatusInternalServerError)
+
+	// Verify that the readiness status follows the health.Ready flag
+	mockStatus.IsReady = false
+	getEventToHandler(t, health.HandleReadiness, readinessPath, http.StatusInternalServerError)
+
 }
 
-// Test The Health Server Via Live HTTP Calls
+// Test The Health Server With Startup Errors
 func TestHealthServer(t *testing.T) {
 
 	logger := logtesting.TestLogger(t).Desugar()
 
 	health := getTestHealthServer()
-	health.Start(logger)
-
-	livenessUri, err := url.Parse(fmt.Sprintf("http://%s:%s%s", testHttpHost, health.HttpPort, livenessPath))
-	assert.Nil(t, err)
-	readinessUri, err := url.Parse(fmt.Sprintf("http://%s:%s%s", testHttpHost, health.HttpPort, readinessPath))
-	assert.Nil(t, err)
-	waitServerReady(readinessUri.String(), 3*time.Second)
-
-	// Test basic functionality - advanced logical tests are in TestHealthHandler
-	getEventToServer(t, livenessUri, http.StatusInternalServerError)
-	getEventToServer(t, readinessUri, http.StatusOK)
-	health.SetAlive(true)
-	getEventToServer(t, livenessUri, http.StatusOK)
-
+	health.HttpPort = "--"
+	err := health.Start(logger)
+	assert.NotNil(t, err)
 	health.Stop(logger)
-
-	// Pause to let async go process finish logging :(
-	// Appears to be race condition between test finishing and logging in health.Stop() above
-	time.Sleep(1 * time.Second)
 }
 
 //

--- a/pkg/channel/distributed/common/health/health_test.go
+++ b/pkg/channel/distributed/common/health/health_test.go
@@ -17,12 +17,13 @@ limitations under the License.
 package health
 
 import (
-	"github.com/stretchr/testify/assert"
 	"io"
-	logtesting "knative.dev/pkg/logging/testing"
 	"net/http"
 	"net/http/httptest"
 	"testing"
+
+	"github.com/stretchr/testify/assert"
+	logtesting "knative.dev/pkg/logging/testing"
 )
 
 // Test Constants

--- a/pkg/channel/distributed/common/health/health_test.go
+++ b/pkg/channel/distributed/common/health/health_test.go
@@ -17,20 +17,16 @@ limitations under the License.
 package health
 
 import (
+	"github.com/stretchr/testify/assert"
 	"io"
+	logtesting "knative.dev/pkg/logging/testing"
 	"net/http"
 	"net/http/httptest"
-	"net/url"
 	"testing"
-	"time"
-
-	"github.com/stretchr/testify/assert"
-	logtesting "knative.dev/pkg/logging/testing"
 )
 
 // Test Constants
 const (
-	testHttpHost  = "localhost"
 	livenessPath  = "/healthz"
 	readinessPath = "/healthy"
 )
@@ -153,32 +149,6 @@ func TestHealthServer(t *testing.T) {
 //
 // Private Utility Functions
 //
-
-// Waits Until A GET Request Succeeds (Or Times Out)
-func waitServerReady(uri string, timeout time.Duration) {
-	// Create An HTTP Client And Send The Request Until Success Or Timeout
-	client := http.DefaultClient
-	for start := time.Now(); time.Since(start) < timeout; {
-		_, err := client.Get(uri) // Don't care what the response actually is, only if there was an error getting it
-		if err == nil {
-			return
-		}
-		time.Sleep(200 * time.Millisecond)
-	}
-}
-
-// Sends A Simple GET Event To A URL Expecting A Specific Response Code
-func getEventToServer(t *testing.T, uri *url.URL, expectedStatus int) {
-
-	// Create An HTTP Client And Send The Request
-	client := http.DefaultClient
-	resp, err := client.Get(uri.String())
-
-	// Verify The Client Response Is As Expected
-	assert.NotNil(t, resp)
-	assert.Nil(t, err)
-	assert.Equal(t, expectedStatus, resp.StatusCode)
-}
 
 // Sends A Request To An HTTP Response Recorder Directly Expecting A Specific Response Code
 func getEventToHandler(t *testing.T, handlerFunc http.HandlerFunc, path string, expectedStatus int) {

--- a/pkg/channel/distributed/common/metrics/stats_reporter.go
+++ b/pkg/channel/distributed/common/metrics/stats_reporter.go
@@ -132,7 +132,7 @@ func (r *Reporter) Report(stats map[string]map[string]interface{}) {
 						return
 					}
 
-					// RecordWrapper The Produced Message Count Metric
+					// Record The Produced Message Count Metric
 					RecordWrapper(ctx, producedMessageCount.M(msgCount))
 
 				} else {

--- a/pkg/channel/distributed/common/metrics/stats_reporter.go
+++ b/pkg/channel/distributed/common/metrics/stats_reporter.go
@@ -87,6 +87,8 @@ func NewStatsReporter(log *zap.Logger) StatsReporter {
 }
 
 // Our RecordWrapper, which defaults to the knative metrics.Record()
+// This wrapper function facilitates minimally-invasive unit testing of the
+// Report functionality without requiring live servers to be started.
 var RecordWrapper = metrics.Record
 
 //

--- a/pkg/channel/distributed/common/metrics/stats_reporter.go
+++ b/pkg/channel/distributed/common/metrics/stats_reporter.go
@@ -86,6 +86,9 @@ func NewStatsReporter(log *zap.Logger) StatsReporter {
 	return &Reporter{logger: log}
 }
 
+// Our RecordWrapper, which defaults to the knative metrics.Record()
+var RecordWrapper = metrics.Record
+
 //
 // Report The Sarama Metrics (go-metrics) Via Knative / OpenCensus Metrics
 //
@@ -129,8 +132,8 @@ func (r *Reporter) Report(stats map[string]map[string]interface{}) {
 						return
 					}
 
-					// Record The Produced Message Count Metric
-					metrics.Record(ctx, producedMessageCount.M(msgCount))
+					// RecordWrapper The Produced Message Count Metric
+					RecordWrapper(ctx, producedMessageCount.M(msgCount))
 
 				} else {
 					r.logger.Warn("Encountered Non Int64 'count' Field In Metric", zap.String("Metric", metricKey))

--- a/pkg/channel/distributed/common/metrics/stats_reporter_test.go
+++ b/pkg/channel/distributed/common/metrics/stats_reporter_test.go
@@ -21,68 +21,26 @@ import (
 	"fmt"
 	"os"
 	"regexp"
-	"strconv"
-	"strings"
 	"testing"
-	"time"
-
-	"io/ioutil"
 
 	"github.com/stretchr/testify/assert"
-	corev1 "k8s.io/api/core/v1"
-	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes/fake"
-	"knative.dev/eventing-kafka/pkg/channel/distributed/common/config"
+	ocstats "go.opencensus.io/stats"
 	"knative.dev/eventing-kafka/pkg/channel/distributed/common/env"
-	commontesting "knative.dev/eventing-kafka/pkg/channel/distributed/common/testing"
-	injectionclient "knative.dev/pkg/client/injection/kube/client"
 	logtesting "knative.dev/pkg/logging/testing"
-	"knative.dev/pkg/metrics"
-	"knative.dev/pkg/system"
 )
 
 // Test The MetricsServer's Report() Functionality
 func TestMetricsServer_Report(t *testing.T) {
-	// see https://github.com/knative-sandbox/eventing-kafka/issues/218
-	t.Skip("Flaky test, therefore skipping for now")
-
 	// Test Data
-	metricsPort := 9878
 	metricsDomain := "eventing-kafka"
 	topicName := "test-topic-name"
 	msgCount := 13579
 
 	// Initialize The Environment For The Test
-	commontesting.SetTestEnvironment(t)
 	assert.Nil(t, os.Setenv(env.MetricsDomainEnvVarKey, metricsDomain))
-
-	// Create An Observability ConfigMap For The InitializeObservability() Call To Watch
-	observabilityConfigMap := &corev1.ConfigMap{
-		TypeMeta: v1.TypeMeta{
-			Kind:       "ConfigMap",
-			APIVersion: corev1.SchemeGroupVersion.String(),
-		},
-		ObjectMeta: v1.ObjectMeta{
-			Name:      metrics.ConfigMapName(),
-			Namespace: system.Namespace(),
-		},
-		Data: map[string]string{
-			"metrics.backend-destination": "prometheus",
-		},
-	}
-
-	// Create The Fake K8S Client And Add It To The ConfigMap
-	fakeK8sClient := fake.NewSimpleClientset(observabilityConfigMap)
-
-	// Add The Fake K8S Client To The Context (Required By InitializeObservability)
-	ctx := context.WithValue(context.TODO(), injectionclient.Key{}, fakeK8sClient)
 
 	// Create A Test Logger
 	logger := logtesting.TestLogger(t).Desugar()
-
-	// Initialize The Observability Watcher
-	err := config.InitializeObservability(ctx, logger.Sugar(), metricsDomain, metricsPort, system.Namespace())
-	assert.Nil(t, err)
 
 	// Create A New StatsReporter To Test
 	statsReporter := NewStatsReporter(logger)
@@ -90,18 +48,20 @@ func TestMetricsServer_Report(t *testing.T) {
 	// Create The Stats / Metrics To Report
 	stats := createTestMetrics(topicName, int64(msgCount))
 
+	measuredMsgCount := 0.0
+
+	RecordWrapperRef := RecordWrapper
+	defer func() { RecordWrapper = RecordWrapperRef }()
+
+	// Set up a mock RecordWrapper() that will capture the produced message count from our stats reporter
+	RecordWrapper = func(ctx context.Context, ms ocstats.Measurement, ros ...ocstats.Options) {
+		measuredMsgCount += ms.Value()
+	}
+
 	// Perform The Test
 	statsReporter.Report(stats)
 
-	// Verify The Results By Querying Metrics Endpoint And Parsing Results
-	resp, err := commontesting.RetryGet(fmt.Sprintf("http://localhost:%v/metrics", metricsPort), 100*time.Millisecond, 20, 404)
-	assert.Nil(t, err)
-	body, err := ioutil.ReadAll(resp.Body)
-	assert.Nil(t, err)
-	err = resp.Body.Close()
-	assert.Nil(t, err)
-	bodyStrings := strings.Split(string(body), "\n")
-	assert.True(t, verifyMetric(bodyStrings, "eventing_kafka_produced_msg_count", topicName, strconv.Itoa(msgCount)))
+	assert.Equal(t, float64(msgCount), measuredMsgCount)
 }
 
 // Utility Function For Creating Sample Test Metrics  (Representative Data From Sarama Metrics Trace - With Custom Test Data)

--- a/pkg/channel/distributed/common/metrics/stats_reporter_test.go
+++ b/pkg/channel/distributed/common/metrics/stats_reporter_test.go
@@ -18,9 +18,7 @@ package metrics
 
 import (
 	"context"
-	"fmt"
 	"os"
-	"regexp"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
@@ -90,26 +88,4 @@ func createTestMetrics(topic string, count int64) map[string]map[string]interfac
 	testMetrics["response-size"] = map[string]interface{}{"75%": 503.25, "95%": 1797, "99%": 1797, "99.9%": 1797, "count": 6, "max": 1797, "mean": 359.5, "median": 72, "min": 72, "stddev": 642.8695435311895}
 	testMetrics["response-size-for-broker-0"] = map[string]interface{}{"75%": 72, "95%": 72, "99%": 72, "99.9%": 72, "count": 5, "max": 72, "mean": 72, "median": 72, "min": 72, "stddev": 0}
 	return testMetrics
-}
-
-// Verifies that the metrics response string slice contains the desired values
-// This is test code so a quick regex check will suffice rather than parsing the Prometheus output in a more structured manner
-func verifyMetric(body []string, name string, topicName string, expectedValue string) bool {
-	for _, line := range body {
-		if isMatch(line, fmt.Sprintf(`^%s`, name)) &&
-			isMatch(line, fmt.Sprintf(`topic="%s"`, topicName)) &&
-			isMatch(line, fmt.Sprintf(` %s$`, expectedValue)) {
-			return true
-		}
-	}
-	return false
-}
-
-// Simple regex match that treats errors as false, for testing only
-func isMatch(source string, regex string) bool {
-	match, err := regexp.MatchString(regex, source)
-	if err != nil {
-		return false
-	}
-	return match
 }

--- a/pkg/channel/distributed/common/metrics/stats_reporter_test.go
+++ b/pkg/channel/distributed/common/metrics/stats_reporter_test.go
@@ -53,7 +53,9 @@ func TestMetricsServer_Report(t *testing.T) {
 
 	// Set up a mock RecordWrapper() that will capture the produced message count from our stats reporter
 	RecordWrapper = func(ctx context.Context, ms ocstats.Measurement, ros ...ocstats.Options) {
-		measuredMsgCount += ms.Value()
+		if ms.Measure().Name() == "produced_msg_count" {
+			measuredMsgCount += ms.Value()
+		}
 	}
 
 	// Perform The Test

